core/tests/protractor_desktop/coreEditorAndPlayerFeatures.js

```diff
@@ -0,0 +1,248 @@
+// Copyright 2019 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview End-to-end tests for the core features of the exploration
+ * editor and player. Core features include the features without which an
+ * exploration cannot be published. These include state content, answer groups,
+ * oppia's feedback and customization_args.
+ */
+
+var forms = require('../protractor_utils/forms.js');
+var general = require('../protractor_utils/general.js');
+var users = require('../protractor_utils/users.js');
+var workflow = require('../protractor_utils/workflow.js');
+
+
+var AdminPage = require('../protractor_utils/AdminPage.js');
+var CreatorDashboardPage =
+  require('../protractor_utils/CreatorDashboardPage.js');
+var ExplorationEditorPage =
+  require('../protractor_utils/ExplorationEditorPage.js');
+var ExplorationPlayerPage =
+  require('../protractor_utils/ExplorationPlayerPage.js');
+var LibraryPage = require('../protractor_utils/LibraryPage.js');
+
+
+describe('Core exploration functionality', function() {
+  var explorationPlayerPage = null;
+  var explorationEditorPage = null;
+  var explorationEditorMainTab = null;
+
+  beforeEach(function() {
```

Comment:
Can we change this (and may be other) `beforeEach` blocks to `beforeAll`? I think that can save us some time. Let me know your thoughts here too!

Reply:
Ah that's something I was thinking about. However, complications might arise in some situations. So the e2e test is supposed to mock a new user trying out that user journey. So conceptually, it makes sense to start the journey from the login/registration for each of them. However, I think for some tests, we can make an exception if the tests are mutually exclusive, and continue to stay that way. But we need to ensure that first before we make this change. 

However, thinking conceptually, I would not want to do this change, and have independent functionality test start afresh. So that we can make sure each individual path functions correctly. 

What's your opinion here? Also, @seanlip, could you weigh in here? Thanks!

Reply:
Also, we can consider using a mix of beforeEach and beforeAll to group up some of the one time setup separately, I think that would be better? [ref](http://breazeal.com/blog/jasmineBefore.html)

Reply:
I agree -- using a combination of beforeEach and beforeAll sounds like a plan. Thanks!

Reply:
Weighing in as requested :)

I think this is a policy decision that the QA / dev workflow teams should decide. I.e. whatever both of you are happy with as being maintainable is fine, as long as there is a clear policy, everything follows that policy, and you update the wiki instructions accordingly.

Reply:
Cool, I would like to discuss this further with Apurv. Also, if it is fine, I would like to defer this optimization to a later PR (or possibly a GSoC project) as it is a largish project. I would like to discuss this sometime soon, and lay out the policy before we offer it as a GSoC project.

Reply:
Let's try to come up with guidelines for this (as to what should be common setup, and what should be omitted from common setup). I can help out. But I'll leave this out of the current PR.